marlowe/test/Spec/Marlowe/Common.hs

```diff
@@ -6,54 +6,332 @@ module Spec.Marlowe.Common where
+    AndObs lhs rhs ->  FalseObs:TrueObs:lhs:rhs:([AndObs x rhs | x <- shrinkObservation lhs]
+                             ++ [AndObs lhs y | y <- shrinkObservation rhs])
+    OrObs lhs rhs ->   FalseObs:TrueObs:lhs:rhs:([OrObs x rhs | x <- shrinkObservation lhs]
+                             ++ [OrObs lhs y | y <- shrinkObservation rhs])
+    NotObs subObs ->   FalseObs:TrueObs:subObs:[NotObs x | x <- shrinkObservation subObs]
+
+
+boundListGenAux :: Int -> Integer -> Gen [Bound]
+boundListGenAux s lb
+  | s > 0 = do inc1 <- simpleIntegerGen
+               inc2 <- simpleIntegerGen
+               let tlb = lb + inc1
+                   thb = tlb + inc2
+               rest <- boundListGenAux (s - 1) thb
+               return (Bound tlb thb:rest)
+  | otherwise = return []
+
+
+listLengthGen :: Gen Int
+listLengthGen = frequency [ (1, return 0)
+                          , (8, return 1)
+                          , (4, return 2)
+                          , (1, return 3)
+                          ]
+
+
+boundListGen :: Gen [Bound]
+boundListGen = do len <- listLengthGen
+                  firstBound <- simpleIntegerGen
+                  boundListGenAux len firstBound
+
+
+actionGenSized :: Int -> Gen Action
+actionGenSized s =
+  oneof [ Deposit <$> accountIdGen <*> partyGen <*> tokenGen <*> valueGenSized (s - 1)
+        , Choice <$> choiceIdGen <*> boundListGen
+        , Notify <$> observationGenSized (s - 1)
+        ]
+
+
+shrinkAction :: Action -> [Action]
+shrinkAction action = case action of
+    Deposit accId party tok val -> Notify FalseObs : [Deposit accId party tok v | v <- shrinkValue val]
+        ++ [Deposit x party tok val | x <- shrinkAccountId accId]
+        ++ [Deposit accId y tok val | y <- shrinkParty party]
+        ++ [Deposit accId party z val | z <- shrinkToken tok]
+    Choice choId boundList -> Notify FalseObs
+        : [Choice choId b | b <- shrinkList (const []) boundList]
+        ++ [Choice x boundList | x <- shrinkChoiceId choId]
+    Notify obs -> [Notify x | x <- shrinkObservation obs]
+
+
+caseRelGenSized :: Int -> Integer -> Gen (Case Contract)
+caseRelGenSized s bn = Case <$> actionGenSized s <*> contractRelGenSized s bn
+
+
+shrinkCase :: Case Contract -> [Case Contract]
+shrinkCase (Case act cont) = [Case act x | x <- shrinkContract cont]
+                              ++ [Case y cont | y <- shrinkAction act]
+
+
+contractRelGenSized :: Int -> Integer -> Gen Contract
+contractRelGenSized s bn
+  | s > 0 = oneof [ return Close
+                  , Pay <$> accountIdGen <*> payeeGen <*> tokenGen
+                        <*> valueGenSized (s `quot` 4)
+                        <*> contractRelGenSized (s - 1) bn
+                  , If <$> observationGenSized (s `quot` 4)
+                       <*> contractRelGenSized (s `quot` 3) bn
+                       <*> contractRelGenSized (s `quot` 3) bn
+                  , Let <$> valueIdGen <*> valueGenSized (s `quot` 4) <*> contractRelGenSized (s `quot` 4) bn
```

Comment:
We should probably change this `contractRelGenSized (s 'quot' 4)` to `contractRelGenSized (s - 1)`, because it only has one subcontract so we should not reduce the size so much

Reply:
I now realise that it was probably me that did that, but still :/

Reply:
Same here, tests just don't terminate in a reasonable time.

Reply:
Which tests?

Reply:
It is fine for shrinking to be slow. Cause if you need to shrink by hand it
is much worse in practice, so you won't mind waiting.